File: NewCode-UpdateCommand/c++/node.cpp
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <sstream>
#include <string>
#include <map>
#include <vector>
#include <stdio.h>   
#include <sys/types.h>   
#include <sys/socket.h>   
#include <netinet/in.h>   
#include <unistd.h>   
#include <errno.h>   
#include <arpa/inet.h>
#include<pthread.h>

using namespace std;
typedef struct node {
    char name[20];
    char ip[50];
    int port;
    int cost;
}node;
typedef struct routing_table {
    char name[20];
    int port;
    node node1;
    node node2;
}routing_table;


//global table
std::map<string,routing_table> global_table_map;

std::vector<std::string> split(const std::string &str, const std::string &delim)
{
    std::vector<std::string> spiltCollection;
    if(str.size()==0)
        return spiltCollection;
    int start = 0;
    int idx = str.find(delim, start);
    while( idx != std::string::npos )
    {
        spiltCollection.push_back(str.substr(start, idx-start));
        start = idx+delim.size();
        idx = str.find(delim, start);
    }
    spiltCollection.push_back(str.substr(start));
    return spiltCollection;
}

void print_table(routing_table table){
    printf(">>> %s Routing table<<<\n",table.name);
    printf("-------------------------------------------------------\n");
    printf("|   destination   |    link cost    |    next hop     |\n");
    printf("|    %-13s|    %-13d|    %-13s|\n",table.node1.name,table.node1.cost,table.node1.name);
    printf("|    %-13s|    %-13d|    %-13s|\n",table.node2.name,table.node2.cost,table.node2.name);
    printf("-------------------------------------------------------\n");
}

bool is_diff(routing_table table1,routing_table table2)
{
    node t1node1 = table1.node1;
    node t1node2 = table1.node2;
    node t2node1 = table2.node1;
    node t2node2 = table2.node2;
    return (t1node1.cost != t2node1.cost || t1node2.cost != t2node2.cost);
}

void print_diff(routing_table table1,routing_table table2)
{
    if(is_diff(table1,table2))
    {
        cout<<"Before Update Table"<<endl;
        print_table(table1);
        cout<<"After Update Table"<<endl;
        print_table(table2);
    }else{
        cout<<"Node "<<table1.name<<" routing table not changed"<<endl;
    }
    node t1node1 = table1.node1;
    node t1node2 = table1.node2;
    node t2node1 = table2.node1;
    node t2node2 = table2.node2;
    if(t1node1.cost != t2node1.cost)
    {
        cout<<"Node "<<t1node1.name<<" cost changed from:"<<t1node1.cost<<" to "<<t2node1.cost<<endl;
    }
    if(t1node2.cost != t2node2.cost)
    {
        cout<<"Node "<<t2node1.name<<" cost changed from:"<<t1node2.cost<<" to "<<t2node2.cost;
    }
}

routing_table read(char* file)
{
    routing_table ret;
    ifstream in(file);
    if (! in.is_open())
    {
        cout<<"Can not open config file "<<file<<endl;
        return ret;
    }
    int line = 0;
    while (!in.eof() )
    {
        char buffer[1024];
        in.getline (buffer,100);
        line++;
        vector<string> array;
        string s = string(buffer);
        array = split(s.c_str(),":");
        if(line == 1)
        {
            sprintf(ret.name,"%s",array[0].c_str());
            ret.port = atoi(array[1].c_str());
        }
        if(line == 2)
        {
            sprintf(ret.node1.name,"%s",array[0].c_str());
            sprintf(ret.node1.ip,"%s",array[1].c_str());
            ret.node1.port = atoi(array[2].c_str());
            ret.node1.cost = atoi(array[3].c_str());
        }
        if(line == 3)
        {
            sprintf(ret.node2.name,"%s",array[0].c_str());
            sprintf(ret.node2.ip,"%s",array[1].c_str());
            ret.node2.port = atoi(array[2].c_str());
            ret.node2.cost = atoi(array[3].c_str());
        }
    }
    return ret;
}

void update_cost(routing_table *table,const char* name,int cost)
{
    int found = 0;
    node node1 = table->node1;
    node node2 = table->node2;
    if(strcmp(node1.name,name) == 0)
    {
        table->node1.cost = cost;
        found = 1;
    }
    if(strcmp(node2.name,name) == 0)
    {
        table->node2.cost = cost;
        found = 1;
    }
    if(!found)
    {
        cout<<"Node not found"<<endl;
    }
}



void sendto(char* ip,int port,routing_table table)
{
    int sock_fd = socket(AF_INET, SOCK_DGRAM, 0);
    struct sockaddr_in addr_serv;
    memset(&addr_serv, 0, sizeof(addr_serv));
    addr_serv.sin_family = AF_INET;
    addr_serv.sin_addr.s_addr = inet_addr(ip);
    addr_serv.sin_port = htons(port);
    int len = sizeof(addr_serv);
    sendto(sock_fd, &table, sizeof(routing_table), 0, (struct sockaddr *)&addr_serv, len);
}

void* server_func(void* arg)
{
    routing_table table;
    memcpy(&table,arg,sizeof(routing_table));
    int sock_fd = socket(AF_INET, SOCK_DGRAM, 0);
    struct sockaddr_in addr_serv;
    int len;
    memset(&addr_serv, 0, sizeof(struct sockaddr_in));
    addr_serv.sin_family = AF_INET; 
    addr_serv.sin_port = htons(table.port);
    addr_serv.sin_addr.s_addr = htonl(INADDR_ANY);
    len = sizeof(addr_serv);
    if(bind(sock_fd, (struct sockaddr *)&addr_serv, sizeof(addr_serv)) < 0)
    {
        perror("bind error:");
        exit(1);
    }
    struct sockaddr_in addr_client;

    while(1)
    {
        routing_table recv;
        char buffer[1024];
        int recv_num = recvfrom(sock_fd, buffer, sizeof(buffer), 0, (struct sockaddr *)&addr_client, (socklen_t *)&len);
        memcpy(&recv,buffer,sizeof(routing_table));
        print_table(recv);
        string name = string(recv.name);
        if(global_table_map.count(name) > 0)
        {
            routing_table old = global_table_map[name];
            print_diff(old,recv);
        }
        global_table_map[name] = recv;
    }
}

int main(int argc,char* argv[])
{
    if(argc != 2)
    {
        cout<<"Useage:"<<"./"<<argv[0]<<" <config file>"<<endl;
        return -1;
    }
    routing_table table = read(argv[1]);
    //server thread
    pthread_t tid;
    int ret = pthread_create(&tid, NULL,server_func,&table);
    while(1)
    {
        cout<<"Input command(load,send,bye,myroutingtable,update):"<<endl;
        string cmd;
        getline(cin,cmd);
        if(cmd == "bye")
            break;
        if(cmd == "send")
        {
            sendto(table.node1.ip,table.node1.port,table);
            sendto(table.node2.ip,table.node2.port,table);
            cout<<"Send routing table finished!"<<endl;

        }else if(cmd == "myroutingtable")
        {
            print_table(table);
        }else if(cmd == "load")
        {
            table = read(argv[1]);
            cout<<"Reload config finished!"<<endl;
        }else if(cmd.find("update") == 0){
            vector<string> cmds = split(cmd.c_str()," ");
            if(cmds.size() != 3)
            {
                cout<<"Useage:update <node name> <cost>"<<endl;
                continue;
            }
            string name = cmds.at(1);
            string cost = cmds.at(2);
            int intCost = atoi(cost.c_str());
            update_cost(&table,name.c_str(),intCost);
            sendto(table.node1.ip,table.node1.port,table);
            sendto(table.node2.ip,table.node2.port,table);
        }else{
            cout<<"Invalid command!"<<endl;
        }
    }
}

File: NewCode-UpdateCommand/python/node.py
import socket
import json
import operator
import sys
import binascii
import struct
import threading
import configparser


def is_diff(obj1,obj2):
    if obj1 == None or obj2 == None:
        return False
    for k in obj1:
        if k == 'node':
            continue
        if obj1[k]['cost'] != obj2[k]['cost']:
            return True
    return False

def print_table(obj):
    if obj == None:
        print(">>>>>> table is empty <<<<<<<<")
        return
    dest1 = obj['link1']['name']
    cost1 = obj['link1']['cost']
    next_hop1 = obj['link1']['name']
    
    dest2 = obj['link2']['name']
    cost2 = obj['link2']['cost']
    next_hop2 = obj['link2']['name']

    print('>>>> ' + obj['node']['name'] + ' routing table <<<<')
    print('-------------------------------------------------------')
    print('|   destination   |    link cost    |    next hop     |')
    print('|    %-13s|    %-13s|    %-13s|' % (dest1,cost1,next_hop1))
    print('|    %-13s|    %-13s|    %-13s|' % (dest2,cost2,next_hop2))
    print('-------------------------------------------------------')


def print_diff(obj1,obj2):
    if obj1 == None or obj2 == None:
        return
    if is_diff(obj1,obj2):
        print("Before Update Table")
        print_table(obj1)
        print("After Update Table")
        print_table(obj2)
        
        for k in obj1:
            if k == "node":
                continue
            if obj2[k]['cost'] != obj1[k]['cost']:
                print("Node " + obj2[k]['name'] + " cost changed from " + str(obj1[k]['cost']) + " to " + str(obj2[k]['cost']))
    else:
        print("Node " + obj1['node']['name'] + " routing table not changed")

def save_table(table):
    global node_configs
    node_configs[table['node']['name']] = table

def listen_thread(port):
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.bind(("0.0.0.0", port))
    while True:
        data, addr = s.recvfrom(1024)
        print(data)

def send(str,ip,port):
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.sendto(bytes(str,'utf8'),(ip,port))

class RecvThread(threading.Thread):
    def __init__(self,port):
        super(RecvThread, self).__init__()
        self.port = port

    def run(self):
        global node_configs
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.bind(("0.0.0.0", self.port))
        while True:
            data, addr = s.recvfrom(1024)
            print("Recv: ");
            dict = json.loads(str(data,encoding='utf-8'))
            print_table(dict)
            name = dict['node']['name']
            old = None
            if name in node_configs:
                old = node_configs[name]
            save_table(dict)
            print_diff(old,dict)
            print("Input command(FirstLoad,FirstSend,Bye,MyRoutingTable,UpdateRouteCost):")

class MyParser(configparser.ConfigParser):
    def as_dict(self):
        d = dict(self._sections)
        for k in d:
            d[k] = dict(d[k])
            d[k].pop('__name__', None)
        return d

if len(sys.argv) != 2:
    print("Useage: python " + sys.argv[0] + " <confg file>")
    sys.exit(-1)

def load_ini(file):
    cf = MyParser()
    cf.read(file)
    return cf.as_dict()

config_dict = load_ini(sys.argv[1])
listen_port = int(config_dict['node']['port'])
#run recv thread
t = RecvThread(int(listen_port))
t.setDaemon(True)
t.start()

s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

node_configs = {}


def send():
    global s
    #send to link1
    s.sendto(bytes(json.dumps(config_dict),'utf8'),(config_dict['link1']['ip'],int(config_dict['link1']['port'])))
    #send to link2
    s.sendto(bytes(json.dumps(config_dict),'utf8'),(config_dict['link2']['ip'],int(config_dict['link2']['port'])))
    print("Send config finished")


def load(ini):
    config_dict = load_ini(ini)
    print("Load config file finished")

def update_cost(node,cost):
    if not cost.isdigit():
        print("Cost is not number")
        return
    found = False
    for k in config_dict:
        if k == "node":
            continue
        v = config_dict[k]
        tmp_name = v['name']
        if tmp_name == name:
            v['cost'] = cost
            found = True
    if not found:
        print("Node <" + name + "> not found in table")
    else:
        send()
while True:
    print("Input command(FirstLoad,FirstSend,Bye,MyRoutingTable,UpdateRouteCost):")
    text = sys.stdin.readline().strip()
    if text == "FirstSend":
        send()
    elif text == "FirstLoad":
        load(sys.argv[1])
    elif text == "Bye":
        break
    elif text == "MyRoutingTable":
        print_table(config_dict)
    elif text.startswith("UpdateRouteCost"):
        cmds = text.split(" ")
        if len(cmds) != 3:
            print("Update command usage:UpdateRouteCost <node name> <cost>")
            continue
        name = cmds[1]
        cost = cmds[2]
        update_cost(name,cost)           
    else:
        print("Invalid command")


